Windows msys2 support

Fabien-Chouteau opened the pull request:
Msys2 is a software distribution platform for Windows (system package manager).

It can provide easy install of external dependencies like we would see on Linux distribution.

With this change msys2 is automatically installed in the `.cache/alire` directory.

Fabien-Chouteau committed:
Add a Setenv procedure in Alire.OS_Lib

This is a wrapper over GNAT.OS_Lib.Setenv with additional logging.

Fabien-Chouteau committed:
Add support for the Pacman system package manager

Fabien-Chouteau committed:
Add support for the Msys2 Windows distro

So far alr detects if Msys2 is either already in the path or installed in standard location.

In the future we can consider installing Msys2 from alr.

Fabien-Chouteau committed:
Remove C from the list of languages in alr.gpr

Fabien-Chouteau committed:
Automatically install msys2 in the .cache folder

Fabien-Chouteau committed:
Move the archive download code to a re-usable function

Fabien-Chouteau committed:
Switch msys2 setup from archive to installer

Fabien-Chouteau committed:
Alire.Origins.Deployers.Source_Archive: capture error output

To avoid unecessary error messages on the tool output

Fabien-Chouteau committed:
testsuite: add a test for msys2 dependency installation

Fabien-Chouteau committed:
Alire.VCSs: remove leading file:// from origin

Some versions of git on Windows do not understand this prefix, and it doesn't seem to be required.

Fabien-Chouteau committed:
Alire.Origins.Deployers.Source_Archive: improve error message

Fabien-Chouteau committed:
Alire.Origins.Deployers.Source_Archive: log tar errors

mosteo commented:
Hi @Fabien-Chouteau , I'm seeing changes about Pacman. Should these be here?

Fabien-Chouteau commented:
> I'm seeing changes about Pacman. Should these be here?

msys2 uses the Pacman project from Arch Linux to install packages.
The code should be reusable on Arch BTW.

mosteo commented:
Ah, that's great then. Thanks for the clarification.

mosteo commented on src/alire/alire-origins-deployers-system-pacman.adb:
```diff
@@ -0,0 +1,118 @@
+                   (Package_Line (Matches (1).First .. Matches (1).Last),
+                    Relaxed => True)); -- Relaxed because some pacman
+                                       --  versions have extra version dash,
+                                       --  e.g. 1.2.3-4
```
If the extra part after the dash does not conform to semantic versioning interpretation we should ignore it instead of parsing it. The thing is that in Semver a version with dash is a pre-release version, so `1.2.3-4 < 1.2.3`

Fabien-Chouteau replied:
Indeed that's not what the hyphen part means in Pacman. I will change that.

mosteo commented on testsuite/tests/index/external-msys2/my_index/index/di/dialog.toml:
```diff
@@ -0,0 +1,10 @@
+[general]
+description = "Just for the test"
+maintainers = ["alejandro@mosteo.com"]
+maintainers-logins = ["mosteo"]
```
Could this confuse us down the road if I didn't write this test?

Fabien-Chouteau replied:
I just copied from another test. We can use `example@example.com` and `example` instead.

mosteo reviewed (commented):
Thanks Fabien, this will surely be a major step. I have only a couple of observations.

Fabien-Chouteau committed:
Alire.Origins.Deployers.System.Pacman: fix version detection

The dash in pacman version is does not correspond to sementic versioning definition.
It is the release number that comes from fixes in the build recipe.

We now ignored it.

Fabien-Chouteau committed:
Alire.Origins.Deployers.System.Pacman: change detection of installed packages

Use the pacman -Qqe command to query installed packages.

Fabien-Chouteau committed:
Add a package to ask questions to the user on the cmd line

Fabien-Chouteau committed:
Alr.Platforms.Windows: query the user before installing msys2

Fabien-Chouteau committed:
Alire.Utils.User_Input: use default if stdin is not a TTY

Fabien-Chouteau committed:
Alr.Platforms.Windows: do not report user's choice as an error

Fabien-Chouteau committed:
Alire.Externals: print message when looking for external

Looking for system package can take some time, it's better to tell the
user that there is something going on.

Fabien-Chouteau committed:
Alr.Platforms.Windows: print message for tools installation

Fabien-Chouteau committed:
Alr.Platforms.Windows: use system deployers

Fabien-Chouteau committed:
Alire.Utils.User_Input: flush input before asking question

Fabien-Chouteau committed:
Alire.Origins.Deployers.System: update user prompt on installation

The user can now:
- Reject installation but continue, can be used to just get sources of
  projects
- Accept all system installations (answer: Always)

mosteo commented:
Our brand new test says some package hash is wrong

Fabien-Chouteau commented:
Check PR doesn't support file renaming:
```
Checking file: 'index/gn/gnat_arm_elf/gnat_arm_elf-10.3.1.toml' (renamed)
Error: Unsupported file status : 'renamed'
Error: Missing 'maintainers-logins' in 'index/gn/gnat_arm_elf/gnat_arm_elf-10.3.1.toml'
```
